server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java

```diff
@@ -85,9 +106,70 @@ private FieldFetcher(List<FieldContext> fieldContexts) {
                 documentFields.put(field, new DocumentField(field, parsedValues));
             }
         }
+        if (includeUnmapped) {
+            collect(documentFields, sourceLookup.loadSourceIfNeeded(), "", 0);
+        }
         return documentFields;
     }
 
+    private void collect(Map<String, DocumentField> documentFields, Map<String, Object> source, String parentPath, int lastState) {
```

Comment:
This turned out to be a bit complex, I'm not sure my idea to do it all in one pass was a good one :) Maybe we could start by re-using the source filtering logic, optimizing later if we see a performance reason ?

Reply:
After looking at what the source filtering code does I actually find this solution less complicated. We are doing less work here than in the source filtering code where whole sub-trees of the source can be filtered out an the exclusion logic more complicated. These two methods that are pulling out the data from the source map while keeping track of where they are in the source tree via the inclusion pattern automaton are not recursive but not that long, and they are implementational details that we don't need to expose. On the other hand, using the source filtering code that was written for another purpose initially on the other hand might lead to problems when that code needs to be altered in the future. Having our "own" logic here doesn't seem to much overhead to me tbh.